File: LLD/Singleton Pattern/java/ThreadSafeSingleton.java
  
     class  Singleton {
        private static Singleton instance = null;
        private static final Object lock = new Object();
        private Singleton(){
            System.out.println("Singleton Constructor called");
        }
        public static Singleton getInstance(){
            if(instance == null)  {
                synchronized (lock) { // Ensures thread safety
                if (instance == null) {
                    instance = new Singleton();
                }
            }
     }
        return instance;
        }
    }

    public class ThreadSafeSingleton {
    public static void main(String[] args) {
        Singleton s1 = Singleton.getInstance();
        Singleton s2 = Singleton.getInstance();
        if (s1 == s2) {
            System.out.println("Both instances are the same!");
        } else {
            System.out.println("Different instances exist!");
        }
    }
}

File: LLD/Singleton Pattern/c++/ThreadSafeSingleton.cpp
#include <iostream>
#include <mutex>
using namespace std;

class Singleton {
    private:
    static Singleton* instance;
    static mutex mtx;
    Singleton(){
       cout << "Singleton Constructor called" << endl;
    }
    public:
    static Singleton* getInstance(){

        if(instance == nullptr){
            lock_guard<mutex> lock(mtx); 
            if(instance == nullptr) { // Double-checked locking
            instance = new Singleton();
            }
        }
        return instance;
     }
};

// Initialize static members
Singleton* Singleton::instance = nullptr;
mutex Singleton::mtx;

int main(){
    Singleton*  s1 = Singleton::getInstance();
    Singleton* s2  = Singleton::getInstance();
    if(s1 == s2){
        cout << "Both instances are the same." << endl;
    } else {
        cout << "Instances are different." << endl;
    }
}